lib/ansible/plugins/connection/local.py

```diff
@@ -131,8 +131,14 @@ def exec_command(self, cmd, in_data=None, sudoable=True):
         return (p.returncode, stdout, stderr)
 
     def _ensure_abs(self, path):
+        ''' join the path with optional cwd and ensure that the type stays the same '''
         if not os.path.isabs(path) and self.cwd is not None:
-            path = os.path.normpath(os.path.join(self.cwd, path))
+            is_binary = isinstance(path, binary_type)
```

Comment:
Why is this needed?

Reply:
Well, the idea is to check the type of the incoming "path" and return the concatenated value in same encoding. In other words, make sure calling _ensure_abs does not unexpectedly change the type of the input parameter if path concatenation is done.

E.g. without this check, the vvv logging in put_file would not look too good, as str path would be converted to bytes by _ensure_abs call.

IMHO the interface of connection plugins (especially LocalConnection) looks bit messy, for example self.cwd is directly modified from outside so there is no control on its encoding.

Anyway, I'm happy to change the implementation if needed. The current version is just the minimal patch to fix the major issue with local playbooks, implemented as safely as I deemed possible.